### DISCUSSION TITLE
Build from code (i.e. not from CLI)

### DISCUSSION THREAD

**Discussion by michaelhays** [tailwindlabs/tailwindcss#11851] [Help] [answered]
I'd like to be able to do something like this to get a string representation of the Tailwind generated styles directly from JavaScript:

```ts
import tailwind from 'tailwindcss'

const style = await tailwind.build({
  input: './src/styles/tailwind.css',
  minify: true,
})
```

However, to my knowledge, there's no way to do this.

Using `execSync` works to some extent:

```ts
import { execSync } from 'node:child_process'

const style = execSync('npx tailwindcss -i ./src/styles/tailwind.css --minify')
```

However, this fails in certain environments with:

```
syscall mkdir
npm ERR! path /home/sbx_user1051
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, mkdir '/home/sbx_user1051'
```

I also tried to call [`build()`](https://github.com/tailwindlabs/tailwindcss/blob/a2fa6932767ab328515f743d6188c2164ad2a5de/src/cli/build/index.js#L8) directly, but ran into similar issues.

**Comment by wongjn** [comment, marked as answer]
You could try using the PostCSS API to programmatically run Tailwind:

```js
const postcss = require('postcss');
const tailwindcss require('tailwindcss');
const baseConfig = require('./tailwind.config');

const sourceCSS = '@tailwind base; @tailwind components; @tailwind utilities';
const config = {
  presets: [baseConfig],
  content: [{ raw: markup }],
};
postcss([tailwindcss(config)])
  .process(sourceCSS)
  .then((css) => {
    // `css` is a string of the compiled CSS.
  });
```

**Reply by michaelhays** [reply]
...was this AI generated? How on earth did you respond so fast?

**Reply by wongjn** [reply]
This a secret between you and me, but I used a [saved reply](https://docs.github.com/en/get-started/writing-on-github/working-with-saved-replies/about-saved-replies).

**Reply by michaelhays** [reply]
Incredible :joy: Thanks so much, this worked great!

**Reply by anthonyLock** [reply]
Hi @wongjn 

Thanks so much for this code! This works amazing for a node server. But when I try it on the front end I get a error 
```
./node_modules/tailwindcss/node_modules/jiti/dist/babel.js:1:99558
Module not found: Can't resolve 'module'
```

Do you know of anyway to get it working on a the front end? 

Many thanks!

**Reply by wongjn** [reply]
You may need to monkey-patch/polyfill Tailwind for use on front-end, of which I don't have any experience with and thus cannot give any concrete details.
